Beta: Multi index search and disjunctive facet search (#915)
meilisearch/instant-meilisearch: .changeset/fresh-tables-clap.md

---
"@meilisearch/instant-meilisearch": minor
---

- Change the behavior of the `or` operator parameter on the [`RefinmentList`](https://www.algolia.com/doc/api-reference/widgets/refinement-list/js/) widget.

⚠️ This impacts the facet distribution number

## Drawback

> Currently, for each `facet value` selected in a different `facet` a separate request is made to Meilisearch. Plus an additional one containing all the facets filters. For example, if I select `Adventure` on genres and `Multiplayer` on players, one request is made with both facet filters, one with only `genres=Adventure` and one with only `players=Multiplayer`

Reviewer: I can't suggest 🤔 

WDYT?
```md
Currently, for each `facet value` selected in a different `facet` a separate request is made to Meilisearch. Plus, an additional one containing all the facets filters. 
For example, if I select `Adventure` on genres and `Multiplayer` on players, one request is made with both facet filters, one with only `genres=Adventure` and one with only `players=Multiplayer`
```

Author: Just to be sure I saw the diff. You only added a coma after `Plus` ?

Reviewer: yes!

> In the next release of Meilisearch, the `multi-index` route should be released (see [PR](https://github.com/meilisearch/meilisearch/pull/3417)). When it is released, the work-around will be removed and only one request will be done in all cases!

Reviewer: Can we promise that?

Author: @gmourier can we?

Reviewer 2: It's a multi-search request route if we want to be more precise, and yes, I'm fairly confident we will be able to do deliver it; we already have a working prototype.

## Explaination

The way the `facet distribution` is calculated changed. The `facet distribution` shows the numbers of hits for each facet. For example:

Given the following facet:

```
Genres
- [ ] Adventure 7
```

The `facet distribution` provides the information that there are 7 hits containing the adventure genre.

> For the example, lets take the following facets:

Reviewer: For the example, let's take the following facets:

---
"@meilisearch/instant-meilisearch": minor
---

- Change the behavior of the `or` operator parameter on the [`RefinmentList`](https://www.algolia.com/doc/api-reference/widgets/refinement-list/js/) widget.

⚠️ This impacts the facet distribution number

## Drawback

> Currently, for each `facet value` selected in a different `facet` a separate request is made to Meilisearch. Plus, an additional one containing all the facets filters.

Reviewer 2: I would have added what it can be tedious to work with, especially regarding bandwidth cost

Author: ```suggestion
Currently, for each `facet value` selected in a different `facet` a separate request is made to Meilisearch. Plus, an additional one containing all the facets filters. It results in tedious search requests and a high bandwidth cost.
```

For example, if I select `Adventure` on genres and `Multiplayer` on players, one request is made with both facet filters, one with only `genres=Adventure` and one with only `players=Multiplayer`

In the next release of Meilisearch, a new `multi-search` API route is planned to be released (see [PR](https://github.com/meilisearch/meilisearch/pull/3417)). When it is released, the work-around will be removed and only one HTTP request will be done in all cases!

> ## Explaination

Reviewer: ```suggestion
## Explanation
```

> The way the `facet distribution` is calculated changed. The `facet distribution` shows the numbers of hits for each facet. For example:

Reviewer 2: ```suggestion
The way the `facetDistribution` is calculated changed. The `facetDistribution` shows the numbers of hits for each facet. For example:
```

Given the following facet:

```
Genres
- [ ] Adventure 7
```

> The `facet distribution` provides the information that there are 7 hits containing the adventure genre.

Reviewer 2: ```suggestion
The `facetDistribution` provides the information that there are `7` hits containing the adventure genre.
```

For the example, let's take the following facets:

```
Genres
- [ ] Adventure 7
- [ ] Action 5
- [ ] Science Fiction 5
Players
- [ ] Multiplayer 11
- [ ] Single player 7
```

Before, when selecting a facet value in a certain facet, the distribution of that facet was impacted.

```
- [x] Adventure 7
- [ ] Action 3 // <- number changed from 5 to 3
- [ ] Science Fiction 1 // <- number changed from 5 to 1
Players
- [ ] Multiplayer 6
- [ ] Single player 3
```

In the new behavior the distribution number are not changed when facet values are selected inside the same facet.

```
Genres
- [x] Adventure 7
- [ ] Action 5 // <- number did not change
- [ ] Science Fiction 5 // <- number did not change
Players
- [ ] Multiplayer 6
- [ ] Single player 3
```

```
Genres
- [x] Adventure 7 // changed because of Multiplayer
- [ ] Action 4
- [ ] Science Fiction 3
Players
- [x] Multiplayer 6
- [ ] Single player 3
```

> This is the conventional way of calculating the facet distribution. If you prefered the old behavior, please consider opening an issue.

Reviewer 2: It could be added that it's the way instant-search is doing it by default. WDYT?

Author: ```suggestion
This is the conventional way of calculating the facet distribution. Similar to Algolia's behavior. If you prefer the old behavior, please consider opening an issue.
```

> This is the conventional way of calculating the facet distribution. If you prefered the old behavior, please consider opening an issue.

Reviewer 2: ```suggestion
This is a more conventional way of calculating the facet distribution. If you prefered the old behavior, please consider opening an issue.
```

> See [complete explaination here](https://github.com/meilisearch/instant-meilisearch/issues/884)

Reviewer: ```suggestion
See [complete explanation here](https://github.com/meilisearch/instant-meilisearch/issues/884)
```
